engine.go

```diff
@@ -495,7 +500,12 @@ type transactionCommittingIter struct {
 }
 
 func (t transactionCommittingIter) Next(ctx *sql.Context) (sql.Row, error) {
-	return t.childIter.Next(ctx)
+	row, err := t.childIter.Next(ctx)
+	// TODO: i know this isn't great, but idk how else :(
```

Comment:
Here's what you do instead:

Introduce a new Node type which will wrap the entire plan if it contains a TriggerExecutor.

Before this new node type begins iterating, it starts a savepoint.

If at any point during iteration it sees an error in its child iterator, it rolls back to that savepoint before propagating the error.